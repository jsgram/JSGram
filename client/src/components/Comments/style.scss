.get-more-comments {
  padding: 1em;
  cursor: pointer;
}

.get-comments {
  color: #5b9bd1;
}

