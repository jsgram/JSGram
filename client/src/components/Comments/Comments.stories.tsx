import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { MemoryRouter } from 'react-router-dom';
import { Comments } from './index';
import './style.scss';

const comments = (username: string, comment: string, _id: string, isEdit: boolean): any => {
    return {
        _id,
        postId: '5d72faed2b49f34231709c53',
        comment,
        authorId: {
            _id: '5d723afacb0c0027cd1e5c94',
            username,
            photoPath: 'https://jsgram-profile-images.s3.eu-west-3.amazonaws.com/1569138592191',
            email: 'dribnucko@gmail.com',
        },
        isEdit,
        newComment: '',
    };
};

const newComment = {
    _id: '5d72faed2b49f3423134340',
    postId: '5d72faed2b49f34231709c53',
    comment: 'Hi',
    authorId: {
        _id: '5d723afacb0c0027cd1e5c94',
        username: 'Karl',
        photoPath: 'https://jsgram-profile-images.s3.eu-west-3.amazonaws.com/1569138592191',
        email: 'dribnucko@gmail.com',
    },
    isEdit: true,
    newComment: 'New comment',
};

export const commentsState = {
    postId: '5d72faed2b49f34231709c53',
    comments: [
        comments('Maks', 'Nice photo', '5d72faed2b49f3423134341', false),
        comments('Bob', 'Like it!', '5d72faed2b49f3423134342', false),
        comments('Amanda', 'Hello', '5d72faed2b49f3423134343', false),
        comments('Winston', 'LIKE', '5d72faed2b49f3423134344', false),
        comments('Frenk', 'Nice photo', '5d72faed2b49f3423134345', false),
    ],
    commentsPage: [{postId: '5d72faed2b49f34231709c53', page: '1'}],
    commentsLoading: false,
    allCommentsLoaded: ['5d72faed2b49f34231709c53'],
    feed: {
        isAdmin: true,
        loggedId: '5d723afacb0c0027cd1e5c94',
        loggedPhotoPath: 'https://jsgram-profile-images.s3.eu-west-3.amazonaws.com/1569138592191',
        loggedUsername: 'didOstap',
    },
};

export const commentsActions = {
    getComments: action('getComments'),
    resetComments: action('resetComments'),
    editCommentAsync: action('editCommentAsync'),
    changeEditStatus: action('changeEditStatus'),
    changeComment: action('changeComment'),
    deleteComment: action('deleteComment'),
    setDefaultCommentToChange: action('setDefaultCommentToChange'),
};

storiesOf('Comments', module)
    .addDecorator((story: any): any => <MemoryRouter>{story()}</MemoryRouter>)
    .add('default comments', () => <Comments {...{...commentsState, ...commentsActions}}/>)
    .add('with "get more comments" button',
        () => <Comments {...{
            ...commentsState,
            allCommentsLoaded: [],
            ...commentsActions,
        }}/>,
    )
    .add('edit comment',
        () => <Comments{...{
            ...commentsState,
            comments: [
                newComment,
                ...commentsState.comments,
            ],
            ...commentsActions,
        }}/>,
    );
