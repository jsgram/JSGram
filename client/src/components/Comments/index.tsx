import React from 'react';
import noAvatar from '../../assets/noAvatar.png';
import { connect } from 'react-redux';
import {
    FIRST_PAGE,
    getComments,
    resetComments,
    deleteComment,
    editCommentAsync,
    changeEditStatus,
    changeComment,
    setDefaultCommentToChange,
} from '../../store/comments/actions';
import { IFeedState } from '../../store/feed/reducers';
import './style.scss';
import { Link } from 'react-router-dom';
import { IAuthor, IComment } from '../../store/comments/reducers';

interface ILocalState {
    postId: string;
    comments: IComment;
    authors: IAuthor;
    allCommentsId: string[];
    commentsPage: any[];
    commentsLoading: boolean;
    allCommentsLoaded: any[];
    feed: IFeedState;
}

interface ICommentInfo {
    _id: string;
    postId: string;
    authorId: string;
    comment: string;
    newComment: string;
    isEdit: boolean;
}

interface IState {
    comments: ILocalState;
    ownProps: {
        postId: string;
    };
    feed: IFeedState;
}

interface IOwnCommentsProps {
    getComments: (commentState: any, commentsLoaded?: boolean) => void;
    resetComments: () => void;
    editCommentAsync: (comment: string, commentId: string) => void;
    changeEditStatus: (commentId: string) => void;
    changeComment: (comment: string, commentId: string) => void;
    deleteComment: (postId: string, authorId: string) => void;
    setDefaultCommentToChange: (postId: string) => void;
}

export type ICommentsProps = IOwnCommentsProps & ILocalState;

export class Comments extends React.Component<ICommentsProps> {
    public componentDidMount(): void {
        if (this.props.postId) {
            this.props.getComments([{postId: this.props.postId, page: FIRST_PAGE}]);
            this.props.setDefaultCommentToChange(this.props.postId);
        }
    }

    public getMoreComments = (): void => {
        const commentStateForCurrentPost =
            this.props.commentsPage.filter((info: { postId: string, page: number }) =>
                info.postId === this.props.postId);

        const commentsLoaded = this.props.allCommentsLoaded.some((post: any) => post === this.props.postId);

        this.props.getComments(commentStateForCurrentPost, commentsLoaded);
    }

    public componentWillUnmount(): void {
        this.props.resetComments();
    }

    public editComment = (comment: string, id: string): void => {
        this.props.editCommentAsync(comment, id);
    }

    public onDeleteComment = (commentId: string, authorId: string): void => {
        this.props.deleteComment(
            commentId, authorId,
        );
    }

    public renderComment = (comment: ICommentInfo): JSX.Element => (
        comment.isEdit ?
            (
                <>
                    <textarea
                        rows={3}
                        className='form-control'
                        value={comment.newComment}
                        onChange={
                            (event: React.ChangeEvent<any>)
                                : void => this.props.changeComment(
                                event.target.value,
                                comment._id,
                            )
                        }
                    />
                    <div className='d-inline float-right edit-delete-comment mt-1'>
                        <i className='fa fa-times fa-lg text-danger mr-2 icon-edit'
                           onClick={(): void => this.props.changeEditStatus(
                               comment._id,
                           )}>
                        </i>
                        <i className='fa fa-check fa-lg text-success icon-edit'
                           onClick={(): void => this.editComment(
                               comment.newComment,
                               comment._id,
                           )}>
                        </i>
                    </div>
                </>
            )
            :
            (
                <>
                    <div className='d-inline float-right edit-delete-comment'>
                        <i
                            className='fa fa-pencil mr-2 edit-comment'
                            onClick={(): void => this.props.changeEditStatus(comment._id)}
                        />
                        <i className='fa fa-trash-o delete-comment' onClick={
                            (): void => this.onDeleteComment(comment._id, comment.authorId)
                        }/>
                    </div>
                    <p>{comment.comment}</p>
                </>
            )
    )

    public renderCommentsTemplate = (commentInfo: ICommentInfo): JSX.Element => {
        const {authorId, comment}: ICommentInfo = commentInfo;
        const {authors, feed: {loggedUsername, isAdmin}}: ICommentsProps = this.props;

        return (
            <div className='one-comment px-3'>
                <div className='d-flex justify-content-between'>
                    <div className='w-100'>
                        <img
                            src={authors[authorId].photoPath || noAvatar}
                            alt='avatar'
                            width={24}
                            height={24}
                            className='img-fluid rounded-circle mt-1 mr-1 mb-1'
                        />
                        <Link to={`/profile/${authors[authorId].username}`}
                              className='text-dark mt-1'
                        >
                            {authors[authorId].username}
                        </Link>
                        {
                            (
                                loggedUsername === authors[authorId].username || isAdmin
                            )
                                ? this.renderComment(commentInfo)
                                :
                                <p>{comment}</p>
                        }
                    </div>
                </div>
            </div>
        );
    }

    public getComments = (): JSX.Element => (
        <div
            className='d-inline float-left get-more-comments'>
            {!this.props.allCommentsLoaded.some((post: any) => post === this.props.postId) &&
            <p
                className='get-comments'
                onClick={this.getMoreComments}
            >
                Get more comments
            </p>
            }
        </div>
    )

    public render(): JSX.Element {
        return (
            <div className='flex-grow-1 comments border-top'>
                {this.props.allCommentsId && this.props.allCommentsId.map((commentId: any) => (
                    <div key={commentId} className='comment'>
                        {
                            this.props.comments[commentId].postId === this.props.postId &&
                            this.renderCommentsTemplate(this.props.comments[commentId])
                        }
                    </div>
                ))}
                {this.getComments()}
            </div>
        );
    }
}

const mapStateToProps = (state: IState, ownProps: { postId: string }): ILocalState => ({
    postId: ownProps.postId,
    comments: state.comments.comments,
    authors: state.comments.authors,
    allCommentsId: state.comments.allCommentsId,
    commentsPage: state.comments.commentsPage,
    commentsLoading: state.comments.commentsLoading,
    allCommentsLoaded: state.comments.allCommentsLoaded,
    feed: state.feed,
});

const mapDispatchToProps = {
    getComments,
    resetComments,
    editCommentAsync,
    changeEditStatus,
    changeComment,
    deleteComment,
    setDefaultCommentToChange,
};

export default connect(mapStateToProps, mapDispatchToProps)(Comments);
