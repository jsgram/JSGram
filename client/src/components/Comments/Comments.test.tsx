import React from 'react';
import { Comments, mapStateToProps } from './index';
import { commentsState } from './Comments.stories';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

describe('Comments component', () => {
    let renderer: any;
    let store: any;
    let getComments: () => string;
    let resetComments: () => string;
    let editCommentAsync: () => string;
    let changeEditStatus: () => string;
    let changeComment: () => string;
    let deleteComment: () => string;
    let setDefaultCommentToChange: () => string;

    beforeEach(() => {
        store = configureStore()(commentsState);

        getComments = jest.fn(() => 'some value');
        resetComments = jest.fn(() => 'some value');
        editCommentAsync = jest.fn(() => 'some value');
        changeEditStatus = jest.fn(() => 'some value');
        changeComment = jest.fn(() => 'some value');
        deleteComment = jest.fn(() => 'some value');
        setDefaultCommentToChange = jest.fn(() => 'some value');

        const actions = {
            getComments,
            resetComments,
            editCommentAsync,
            changeEditStatus,
            changeComment,
            deleteComment,
            setDefaultCommentToChange,
        };

        renderer = shallow(<Comments {...{...commentsState, ...actions}}/>);
    });

    test('componentDidMount - success', () => {
        renderer.instance().componentDidMount();
        expect(getComments).toHaveReturnedWith('some value');
        expect(setDefaultCommentToChange).toHaveReturnedWith('some value');
    });

    test('componentWillUnmount - success', () => {
        renderer.instance().componentWillUnmount();
        expect(resetComments).toHaveReturnedWith('some value');
    });

    test('getMoreComments - success', () => {
        renderer.instance().getMoreComments();
        expect(getComments).toHaveReturnedWith('some value');
    });

    test('edit - success', () => {
        renderer.instance().editComment('some value', '1c');
        expect(editCommentAsync).toHaveReturnedWith('some value');
    });

    test('delete - success', () => {
        renderer.instance().onDeleteComment('1c', '1c');
        expect(deleteComment).toHaveReturnedWith('some value');
    });

    test('render - success', () => {
        expect(renderer).toMatchSnapshot();
    });

    test('render with redux - success', () => {
        const props = {
            getComments,
            resetComments,
            editCommentAsync,
            changeEditStatus,
            changeComment,
            deleteComment,
            setDefaultCommentToChange,
        };

        renderer = shallow(
            <Provider store={store}>
                <Comments {...{...commentsState, ...props}}/>
            </Provider>,
        );

        expect(renderer).toMatchSnapshot();
    });

    test('mapStateToPropsSuccess - success', () => {
        expect(mapStateToProps({
            comments: commentsState,
            feed: commentsState.feed,
        }, {postId: commentsState.postId})).toEqual(commentsState);
    });
});
