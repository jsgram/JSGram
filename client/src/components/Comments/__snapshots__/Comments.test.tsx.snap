// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments component render - success 1`] = `ShallowWrapper {}`;

exports[`Comments component render with redux - success 1`] = `ShallowWrapper {}`;
